File: RRFFilesManager/Intake/PotentialClientInfo.vb
Public Class PotentialClientInfo
    Public Property IntakeForm As Intake

    Public ReadOnly Property PreliminaryInfo() As PreliminaryInfo
        Get
            Return IntakeForm.PreliminaryInfo
        End Get
    End Property

    Public ReadOnly Property PCIDateOfBirth() As Date?
        Get
            If (String.IsNullOrWhiteSpace(Me.monthBirth.Text) Or String.IsNullOrWhiteSpace(Me.dayBirth.Text) Or String.IsNullOrWhiteSpace(Me.yearBirth.Text)) Then
                Return Nothing
            End If
            Return Date.Parse(Me.monthBirth.Text + "-" + dayBirth.Text + "-" + yearBirth.Text)
        End Get
    End Property

    Public ReadOnly Property newLimDate() As Date?
        Get
            If (String.IsNullOrWhiteSpace(Me.monthBirth.Text) Or String.IsNullOrWhiteSpace(Me.dayBirth.Text) Or String.IsNullOrWhiteSpace(Me.yearBirth.Text)) Then
                Return Nothing
            End If
            Dim newYear As Integer = yearBirth.Text
            newYear = newYear + 20
            Return Date.Parse(Me.monthBirth.Text + "-" + dayBirth.Text + "-" + newYear.ToString())
        End Get
    End Property

    Private Sub PCIMobileNumber_TextChanged_1(sender As Object, e As EventArgs) Handles PCIMobileNumber.TextChanged
        Dim newPN As String
        If PCIMobileCarrier.Text <> "Other" Then
            Dim dr() As System.Data.DataRow
            dr = ActionLogDBDataSet.Tables("MobileCarrier").Select("Carrier='" & PCIMobileCarrier.Text & "'")
            If dr.Length > 0 Then
                Dim gateString As String = dr(0)("Gate").ToString()
                newPN = PCIMobileNumber.Text.Replace("(", "")
                newPN = newPN.Replace(")", "")
                newPN = newPN.Replace(" ", "")
                newPN = Trim(newPN.Replace("-", ""))
                Me.PCIEmailToText.Text = newPN + "@" + gateString
            End If
        Else
            Me.PCIEmailToText.Text = ""
        End If
    End Sub

    Private Sub monthBirth_SelectedIndexChanged(sender As Object, e As EventArgs) Handles monthBirth.SelectedIndexChanged
        If Me.yearBirth.MaskCompleted = False Or monthBirth.Text = "" Or dayBirth.Text = "" Then
            If Me.monthBirth.Text = "" Then Exit Sub
            If Me.dayBirth.Text = "" Then Exit Sub
            If Me.yearBirth.MaskCompleted = False Then Exit Sub
        Else
            Dim date1 As Date = PreliminaryInfo.DateOfLossDateTimePicker.Value
            Dim date2 As Date = PCIDateOfBirth.Value
            Dim age As Long = DateDiff(DateInterval.Year, date2, date1)
            If age < 18 Then
                PreliminaryInfo.LimitationPeriodTextBox.Text = Format(Me.newLimDate.Value, "MMM-dd-yyyy")
                MsgBox("Limitation Period has been changed to " + PreliminaryInfo.LimitationPeriodTextBox.Text)
            End If
        End If
    End Sub

    Private Sub dayBirth_TextChanged(sender As Object, e As EventArgs) Handles dayBirth.TextChanged
        If Me.yearBirth.MaskCompleted = False Or monthBirth.Text = "" Or dayBirth.Text = "" Then
            If Me.monthBirth.Text = "" Then Exit Sub
            If Me.dayBirth.Text = "" Then Exit Sub
            If Me.yearBirth.MaskCompleted = False Then Exit Sub
        Else
            Dim date1 As Date = PreliminaryInfo.DateOfLossDateTimePicker.Value
            Dim date2 As Date = PCIDateOfBirth.Value
            Dim age As Long = DateDiff(DateInterval.Year, date2, date1)
            If age < 18 Then
                PreliminaryInfo.LimitationPeriodTextBox.Text = Format(Me.newLimDate.Value, "MMM-dd-yyyy")
                MsgBox("Limitation Period has been changed to " + PreliminaryInfo.LimitationPeriodTextBox.Text)
            End If
        End If
    End Sub

    Private Sub yearBirth_TextChanged(sender As Object, e As EventArgs) Handles yearBirth.TextChanged
        If Me.yearBirth.MaskCompleted = False Or monthBirth.Text = "" Or dayBirth.Text = "" Then
            If Me.monthBirth.Text = "" Then Exit Sub
            If Me.dayBirth.Text = "" Then Exit Sub
            If Me.yearBirth.MaskCompleted = False Then Exit Sub
        Else
            Dim date1 As Date = PreliminaryInfo.DateOfLossDateTimePicker.Value
            Dim date2 As Date = PCIDateOfBirth.Value
            Dim age As Long = DateDiff(DateInterval.Year, date2, date1)
            If age < 18 Then
                PreliminaryInfo.LimitationPeriodTextBox.Text = Format(Me.newLimDate.Value, "MMM-dd-yyyy")
                MsgBox("Limitation Period has been changed to " + PreliminaryInfo.LimitationPeriodTextBox.Text)
            End If
        End If
    End Sub
    Private Sub PCIMobileCarrier_SelectedIndexChanged(sender As Object, e As EventArgs) Handles PCIMobileCarrier.SelectedIndexChanged
        Dim newPN As String
        If PCIMobileCarrier.Text <> "Other" Then
            Dim dr() As System.Data.DataRow
            dr = ActionLogDBDataSet.Tables("MobileCarrier").Select("Carrier='" & PCIMobileCarrier.Text & "'")
            If dr.Length > 0 Then
                Dim gateString As String = dr(0)("Gate").ToString()
                newPN = PCIMobileNumber.Text.Replace("(", "")
                newPN = newPN.Replace(")", "")
                newPN = newPN.Replace(" ", "")
                newPN = Trim(newPN.Replace("-", ""))
                Me.PCIEmailToText.Text = newPN + "@" + gateString
            End If
        Else
            Me.PCIEmailToText.Text = ""
        End If
    End Sub

    'Private Sub PCIMobileNumber_TextChanged(sender As Object, e As EventArgs)
    '    Dim newPN As String
    '    If PCIMobileCarrier.Text <> "Other" Then
    '        Dim dr() As System.Data.DataRow
    '        dr = ActionLogDBDataSet.Tables("MobileCarrier").Select("Carrier='" & PCIMobileCarrier.Text & "'")
    '        If dr.Length > 0 Then
    '            Dim gateString As String = dr(0)("Gate").ToString()
    '            newPN = PCIMobileNumber.Text.Replace("(", "")
    '            newPN = newPN.Replace(")", "")
    '            newPN = newPN.Replace(" ", "")
    '            newPN = Trim(newPN.Replace("-", ""))
    '            Me.PCIEmailToText.Text = newPN + "@" + gateString
    '        End If
    '    Else
    '        Me.PCIEmailToText.Text = ""
    '    End If
    'End Sub
    Private Sub PotentialClientInfo_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        MobileCarrierTableAdapter.Fill(ActionLogDBDataSet.MobileCarrier)
        ProvincesTableAdapter.Fill(ActionLogDBDataSet.Provinces)
        IntakesTableAdapter.Fill(ActionLogDBDataSet.Intakes)
        Dim obj = IntakesBindingSource.AddNew()
        yearBirth.Text = "1970"
    End Sub

    Public Function Validate() As Boolean
        If Me.PCISalutation.Text = "" Then MsgBox("Please enter Salutation") : Return False
        If Me.PCIFirstName.Text = "" Then MsgBox("Please enter First Name") : Return False
        If Me.PCILastName.Text = "" Then MsgBox("Please enter Last Name") : Return False
        If Me.PCIEmail.Text = "" And Me.PCIAddress.Text = "" And Me.PCIPostalCode.Text = "" And Me.PCICity.Text = "" And Me.PCIProvince.Text = "" Then MsgBox("Please enter Email or Address") : Return False
        If Me.PCIEmail.Text = "" Then
            If Me.PCIAddress.Text = "" Or Me.PCIPostalCode.Text = "" Or Me.PCICity.Text = "" Or Me.PCIProvince.Text = "" Then
                If Me.PCIAddress.Text = "" Then MsgBox("Please enter Address") : Return False
                If Me.PCICity.Text = "" Then MsgBox("Please enter City") : Return False
                If Me.PCIProvince.Text = "" Then MsgBox("Please enter Province") : Return False
                If Me.PCIPostalCode.Text = "" Then MsgBox("Please enter Postal Code") : Return False
            End If
        End If
        If Me.PCIHomeNumber.MaskCompleted = False And Me.PCIWorkNumber.MaskCompleted = False And Me.PCIMobileNumber.MaskCompleted = False Then MsgBox("Please enter a Phone Number") : Return False
        If Me.yearBirth.MaskCompleted = False Or monthBirth.Text = "" Or dayBirth.Text = "" Then
            If Me.monthBirth.Text = "" Then MsgBox("Please enter Month of Birth") : Return False
            If Me.dayBirth.Text = "" Then MsgBox("Please enter Day of Birth") : Return False
            If Me.yearBirth.MaskCompleted = False Then MsgBox("Please enter Year of Birth") : Return False

        End If

        Return True
    End Function

    Public Sub OnNext()
        'IntakeForm.IntakesBindingSource.EndEdit()
        'IntakeForm.IntakesTableAdapter.Update(ActionLogDBDataSet.Intakes)
    End Sub
End Class

File: RRFFilesManager/IntakeForm/PotentialClientInfo.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using RRFFilesManager.DataAccess;
using RRFFilesManager.Abstractions;
using System.Text.RegularExpressions;

namespace RRFFilesManager.IntakeForm
{
    public partial class PotentialClientInfo : UserControl
    {
        public PotentialClientInfo()
        {
            InitializeComponent();
        }

        private static PotentialClientInfo instance;
        public static PotentialClientInfo Instance => instance ?? (instance = new PotentialClientInfo());
        public void SetClient(Client client) {
            IntakeForm.Intake.Client = client;
            FillForm(client);
        }
        public DateTime? PCIDateOfBirth
        {
            get
            {
                if (string.IsNullOrWhiteSpace(MonthBirth.Text) || string.IsNullOrWhiteSpace(DayBirth.Text) || string.IsNullOrWhiteSpace(YearBirth.Text))
                {
                    return default;
                }

                return DateTime.Parse($"{MonthBirth.Text}-{DayBirth.Text}-{YearBirth.Text}");
            }
        }

        private void PCIMobileNumber_TextChanged_1(object sender, EventArgs e)
        {

        }

        private void monthBirth_SelectedIndexChanged(object sender, EventArgs e)
        {

        }

        private void dayBirth_TextChanged(object sender, EventArgs e)
        {

        }

        private void yearBirth_TextChanged(object sender, EventArgs e)
        {

        }

        private void PCIMobileCarrier_SelectedIndexChanged(object sender, EventArgs e)
        {
            PCIMobileNumber_TextChanged(sender, e);
        }

        private void PCIMobileNumber_TextChanged(object sender, EventArgs e)
        {
            var mobileCarrier = (MobileCarrier)PCIMobileCarrier.SelectedItem;
            if (mobileCarrier != null)
                PCIEmailToText.Text = $"{Regex.Replace(PCIMobileNumber.Text, "[ ()-]", "")}@{mobileCarrier?.Gate.Trim()}";
            else
                PCIEmailToText.Text = "";
        }
        private void PotentialClientInfo_Load(object sender, EventArgs e)
        {
                PCIProvince.DataSource = Program.DBContext.Provinces.ToList();
                PCIProvince.DisplayMember = nameof(Province.Description);

                PCIMobileCarrier.DataSource = Program.DBContext.MobileCarriers.ToList();
                PCIMobileCarrier.DisplayMember = nameof(MobileCarrier.Description);
            
            YearBirth.Text = "1970";
        }

        public new bool Validate()
        {
            if (string.IsNullOrEmpty(this.PCISalutation.Text))
            {
                MessageBox.Show("Please enter Salutation");
                return false;
            }

            if (string.IsNullOrEmpty(this.PCIFirstName.Text))
            {
                MessageBox.Show("Please enter First Name");
                return false;
            }

            if (string.IsNullOrEmpty(this.PCILastName.Text))
            {
                MessageBox.Show("Please enter Last Name");
                return false;
            }

            if (string.IsNullOrEmpty(this.PCIEmail.Text) & string.IsNullOrEmpty(this.PCIAddress.Text) & string.IsNullOrEmpty(this.PCIPostalCode.Text) & string.IsNullOrEmpty(this.PCICity.Text) & string.IsNullOrEmpty(this.PCIProvince.Text))
            {
                MessageBox.Show("Please enter Email or Address");
                return false;
            }

            if (string.IsNullOrEmpty(this.PCIEmail.Text))
            {
                if (string.IsNullOrEmpty(this.PCIAddress.Text) | string.IsNullOrEmpty(this.PCIPostalCode.Text) | string.IsNullOrEmpty(this.PCICity.Text) | string.IsNullOrEmpty(this.PCIProvince.Text))
                {
                    if (string.IsNullOrEmpty(this.PCIAddress.Text))
                    {
                        MessageBox.Show("Please enter Address");
                        return false;
                    }

                    if (string.IsNullOrEmpty(this.PCICity.Text))
                    {
                        MessageBox.Show("Please enter City");
                        return false;
                    }

                    if (string.IsNullOrEmpty(this.PCIProvince.Text))
                    {
                        MessageBox.Show("Please enter Province");
                        return false;
                    }

                    if (string.IsNullOrEmpty(this.PCIPostalCode.Text))
                    {
                        MessageBox.Show("Please enter Postal Code");
                        return false;
                    }
                }
            }

            if (this.PCIHomeNumber.MaskCompleted == false & this.PCIWorkNumber.MaskCompleted == false & this.PCIMobileNumber.MaskCompleted == false)
            {
                MessageBox.Show("Please enter a Phone Number");
                return false;
            }

            if (this.YearBirth.MaskCompleted == false | string.IsNullOrEmpty(this.MonthBirth.Text) | string.IsNullOrEmpty(this.DayBirth.Text))
            {
                if (string.IsNullOrEmpty(this.MonthBirth.Text))
                {
                    MessageBox.Show("Please enter Month of Birth");
                    return false;
                }

                if (string.IsNullOrEmpty(this.DayBirth.Text))
                {
                    MessageBox.Show("Please enter Day of Birth");
                    return false;
                }

                if (this.YearBirth.MaskCompleted == false)
                {
                    MessageBox.Show("Please enter Year of Birth");
                    return false;
                }
            }

            return true;
        }

        public void OnNext()
        {
            FillOrCreateIntakeClient();
        }

        public void FillClient(Client client)
        {
            client.Salutation = PCISalutation.Text;
            client.FirstName = PCIFirstName.Text;
            client.LastName = PCILastName.Text;
            client.Address = PCIAddress.Text;
            client.SuiteApt = PCISuiteApt.Text;
            client.Email = PCIEmail.Text;
            client.Province = (Province)PCIProvince.SelectedItem;
            client.City = PCICity.Text;
            client.PostalCode = PCIPostalCode.Text;
            client.HomeNumber = PCIHomeNumber.Text;
            client.WorkNumber = PCIWorkNumber.Text;
            client.MobileNumber = PCIMobileNumber.Text;
            client.MobileCarrier = PCIMobileCarrier.Text;
            client.EmailToText = PCIEmailToText.Text;
            client.DateOfBirth = PCIDateOfBirth;
            client.OtherNotes = PCIOtherNotes.Text;
        }

        public void FillForm(Client client)
        {
            PCISalutation.Text = client.Salutation;
            PCIFirstName.Text = client.FirstName;
            PCILastName.Text = client.LastName;
            PCIAddress.Text = client.Address;
            PCISuiteApt.Text = client.SuiteApt;
            PCIEmail.Text = client.Email;
            PCIProvince.SelectedItem = client.Province;
            PCICity.Text = client.City;
            PCIPostalCode.Text = client.PostalCode;
            PCIHomeNumber.Text = client.HomeNumber;
            PCIWorkNumber.Text = client.WorkNumber;
            PCIMobileNumber.Text = client.MobileNumber;
            PCIMobileCarrier.Text = client.MobileCarrier;
            PCIEmailToText.Text = client.EmailToText;
            YearBirth.Text = client.DateOfBirth?.Year.ToString();
            MonthBirth.Text = client.DateOfBirth?.Month.ToString();
            DayBirth.Text = client.DateOfBirth?.Day.ToString();
            PCIOtherNotes.Text = client.OtherNotes;
        }

        public Client CreateClient()
        {
                var client = new Client();
                FillClient(client);
                Program.DBContext.Clients.Add(client);
                Program.DBContext.SaveChanges();
                return client;
        }

        public Client UpdateClient(int clientId)
        {
                var client = Program.DBContext.Clients.FirstOrDefault(s => s.ID == clientId);
                FillClient(client);
                Program.DBContext.SaveChanges();
                return client;
        }

        public void FillOrCreateIntakeClient()
        {
            if (IntakeForm.Intake.Client == null)
                IntakeForm.Intake.Client = CreateClient();
            else
                IntakeForm.Intake.Client = UpdateClient(IntakeForm.Intake.Client.ID);
        }

        private void PCISalutation_SelectedIndexChanged(object sender, EventArgs e)
        {

        }

        private void PCIProvince_SelectedIndexChanged(object sender, EventArgs e)
        {

        }

        private void FindClientButton_Click(object sender, EventArgs e)
        {
            FindClient.Instance.Show();
        }
    }
}
